### PR TITLE
Deprecate `@` based syntax for accessing iceberg snapshots

### PR DESCRIPTION
**Pull request by phd3**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

this deprecates `@` based syntax for accessing iceberg snapshots.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

iceberg connector

> How would you describe this change to a non-technical end user or system administrator?

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

() No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Iceberg
* Deprecate Iceberg snapshot access using mangled table names, in favor of the `AS OF` syntax. The old behavior can be restored by `allow_legacy_snapshot_syntax`  session property or `iceberg.allow-legacy-snapshot-syntax` config property. ({issue}`10768`)
```

### CODE REVIEW

**Comment by b-slim** [comment]
Thanks, @phd3  for the patch I am wondering why shall we keep around this ambiguous access, I think it is better to fail when a snapshot id is not valid and the user can still access the latest snapshot when using table name with no snapshot. IOW this flag is yet another flag with no clear added value?

**Comment by phd3** [comment]
> I think it is better to fail when a snapshot id is not valid

Current behavior also allows interpreting that number as a TS. Since this syntax has been around for a while, IMO would be worth providing a switch before fully removing. but let's see what others think too

**Comment by phd3** [comment]
Flaky #10752

**Comment by martint** [comment]
`AS OF` syntax is already merged into Trino, BTW.

**Review by electrum — APPROVED** [review]

**Comment by electrum** [comment]
Should we make this config an enum of BOTH, SNAPSHOT_ONLY, NONE?

**Comment by phd3** [comment]
@electrum updated to add enums. would you like to take another look?

**Comment by phd3** [comment]
@findepi @electrum is allow/deny sufficient now since #10258 merged? or still want to keep enum?

**Comment by findepi** [comment]
@phd3  i am fine with simple boolean flag

**Comment by phd3** [comment]
@findepi ptal

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSessionProperties.java [hunk 1]
```diff
@@ -245,6 +246,10 @@ public IcebergSessionProperties(
                         "Minimal retention period for remove_orphan_files procedure",
                         icebergConfig.getRemoveOrphanFilesMinRetention(),
                         false))
+                .add(booleanProperty(ALLOW_LEGACY_SNAPSHOT_SYNTAX,
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSessionProperties.java:249]: nit: line break before ALLOW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java [hunk 2]
```diff
@@ -218,8 +218,17 @@ public static Map<String, Object> getIcebergTableProperties(Table icebergTable)
         return properties.buildOrThrow();
     }
 
-    public static long resolveSnapshotId(Table table, long snapshotId)
+    public static long resolveSnapshotId(Table table, long snapshotId, boolean allowLegacySnapshotSyntax)
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:221]: deprecate the method

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java [hunk 3]
```diff
@@ -228,7 +237,7 @@ public static long resolveSnapshotId(Table table, long snapshotId)
                 .filter(entry -> entry.timestampMillis() <= snapshotId)
                 .map(HistoryEntry::snapshotId)
                 .findFirst()
-                .orElseThrow(() -> new TrinoException(ICEBERG_INVALID_SNAPSHOT_ID, format("Invalid snapshot [%s] for table: %s", snapshotId, table)));
+                .orElseThrow(() -> new TrinoException(ICEBERG_INVALID_SNAPSHOT_ID, format("Invalid snapshot [%s] for table: %s", snapshotId, table.name())));
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:240]: nit: unrelated

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergSnapshots.java [hunk 4]
```diff
@@ -44,8 +48,40 @@ public void testReadingFromSpecificSnapshot()
...
+        assertQuery(sessionWithLegacySyntaxSupport(), selectAllFromLatestId, "VALUES(1, 1)");
+        assertQuery(sessionWithLegacySyntaxSupport(), selectFromPartitionsTable, "VALUES(1)");
+
+        // DISABLED
+        assertThatThrownBy(() -> query(getSession(), selectAllFromFutureTimeStamp))
+                .hasMessageContaining(notSupportedError(tableName, futureTimeStamp));
+        assertThatThrownBy(() -> query(getSession(), selectAllFromLatestId))
+                .hasMessageContaining(notSupportedError(tableName, getLast(ids)));
+        assertThatThrownBy(() -> query(getSession(), selectFromPartitionsTable))
+                .hasMessageContaining(notSupportedError(tableName, getLast(ids)));
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergSnapshots.java]: hasMessageMatching and then you can inline the expected message pattern

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergSnapshots.java [hunk 5]
```diff
@@ -74,4 +111,20 @@ private List<Long> getSnapshotsIdsByCreationOrder(String tableName)
...
+    {
+        return Session.builder(getSession())
+                .setCatalogSessionProperty("iceberg", "allow_legacy_snapshot_syntax", "true")
+                .build();
+    }
+
+    private String notSupportedError(String tableName, long id)
+    {
+        return format(
+                "Failed to access snapshot %s for table %s.\"%s\". This pattern for accessing Iceberg tables is not supported.",
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergSnapshots.java]: I think the message should mention a) the unsupported syntax that was used and b) direct users to use AS OF syntax, or enable compatibility toggle to temporarily restore previous behavior.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestLegacySyntaxForIcebergSnapshotAccess.java [hunk 6]
```diff
@@ -28,7 +28,7 @@
 import static java.util.stream.Collectors.toList;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
-public class TestIcebergSnapshots
+public class TestLegacySyntaxForIcebergSnapshotAccess
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestLegacySyntaxForIcebergSnapshotAccess.java:31]: You can ssquash this together with `Allow configuring support for snapshot access with legacy syntax` into a commit name sth like `Deprecate Iceberg snapshot access using mangled table names`.

**Review by findepi — APPROVED** [review]
